feat(checklist): initial implementation of components

capJavert opened the pull request:
## Changes

### Describe what this PR does

This PR implements:
- basic squad checklist with dynamic steps based on role
- adds current copy to steps from design
- creates step components for specific steps (missing logic and completion logic, wanted to split those up in sub tickets since this PR does enough already)
- integrates `useActions` hook to build steps for checklist

Implementation detail of checklist components:
- Create checklist components that will be used to display and assemble squad checklists needed for v5
- Idea is to have the base components that can be used and composed into specific implementation to satisfy different needs for checklist in scope of v5 and in the future
- `ChecklistCard` is the base component that accepts steps and creates a card
  - each step contains an action and additional props that are needed to render it (eg. additional component, classes etc.)
  - Checklist card can also composed into parent component to have its own specific logic on how to map actions to steps
  - see `SquadChecklistCard` - this one contains all the logic to map steps and actions per current member role
  - in the future when we have different checklist for some other parts of the app we can follow the same pattern
- `ChecklistStep` is a default component that is used to render steps inside the checklist
  - it can be composed into new components with additional views that are needed for some checklist step or some generic ones to be reused across different checklists (eg. install extension step)
  - see `InstallExtensionChecklistStep` and other `*ChecklistStep` components

You can see all of this in action on squad page.

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

### **Please make sure existing components are not breaking/affected by this PR**

## Manual Testing

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-1293 #done

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,90 @@
+  steps,
+  onRequestClose,
+}: ChecklistCardProps): ReactElement => {
+  const activeStep = useMemo(
+    () => steps.find((item) => !item.action.dateCompleted)?.action.type,
+    [steps],
+  );
+  const [checkedStep, setCheckedStep] = useState<string>(activeStep);
+  const isDone = useMemo(
+    () => steps.every((item) => !!item.action.dateCompleted),
+    [steps],
+  );
```
This almost sounds like a perfect fit for a hook, especially since a lot of this will be controlled by external placed actions.
What do you think?

capJavert replied:
Yes, as written in description, some of the logic should go to useActions hook, its here until we see how that one turns out.

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,90 @@
+
+  return (
+    <div className={className}>
+      <Card className="p-0 rounded-14 border-cabbage-40 hover:!border-cabbage-40 w-[340px]">
```
I don't think the color options here are good, please refrain from using colors directly as it doesn't include light/dark mode.
There should be a named version of this available. (Let me know if you can't find it)

capJavert replied:
Yeah got it, It is clunky because in design colors are written like this and then finding those in theme is hell and they often mismatch 💦 . I'll try and remap to theme colors for sure and let you know.

<img width="304" alt="image" src="https://user-images.githubusercontent.com/9803078/235086572-5dde04fd-73cc-4d2c-a238-8f07c5fa857e.png">
<img width="299" alt="image" src="https://user-images.githubusercontent.com/9803078/235086630-732ce1e5-7294-4efa-9210-388cb74e74bc.png">

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,90 @@
+          <p className="mb-1 font-bold typo-body text-theme-label-primary">
+            {title}
+          </p>
+          <p className="text-salt-90 typo-callout">{description}</p>
```
Also here the color can't be right.

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,90 @@
+          <p className="text-salt-90 typo-callout">{description}</p>
+          {typeof onRequestClose === 'function' && (
+            <button
+              className="absolute top-4 right-4 text-salt-90"
```
Same note

capJavert replied:
@rebelchris so for this the right value should be `text-theme-color-salt`?

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,90 @@
+                active={activeStep === step.action.type}
+                onToggle={onToggleStep}
+                className={{
+                  checkmark: isDone && 'text-cabbage-40',
```
Here as well color issue

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,94 @@
+          <QuaternaryButton
+            className={classNames(
+              '-ml-2 text-salt-90',
+              isCompleted && 'text-pepper-10',
```
Color is not right here.

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,94 @@
+                  active && 'p-1 rounded-full bg-theme-bg-cabbage-opacity-24',
+                )}
+              >
+                <ChecklistAIcon
```
What is Checklist A? (not sure the name is a typo or not?)

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,94 @@
+              >
+                <ChecklistAIcon
+                  className={classNames(
+                    active && 'text-cabbage-40',
```
Color issue

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,94 @@
+              className={classNames(
+                'typo-callout',
+                active ? 'font-bold text-theme-label-primary' : 'font-normal',
+                isCompleted && 'text-pepper-10',
```
Color

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,94 @@
+        </div>
+        <ArrowIcon
+          className={classNames(
+            !isOpen && 'rotate-180 text-salt-90',
```
Color

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,94 @@
+        <div className="my-2 ml-9">
+          <p
+            className={classNames(
+              'text-salt-90 typo-callout',
```
Color

rebelchris commented on packages/shared/src/components/checklist/NewSquadAndSiteMemberChecklistCard.tsx:
```diff
@@ -0,0 +1,109 @@
+import BellIcon from '../icons/Bell';
+import { FlexRow } from '../utilities';
+
+const NewSquadAndSiteMemberChecklistCard = ({
```
I'm not 100% sure this format is correct, in the latest spec we decided that steps are fully optional based on role + status.
So someone could be a multiple or roles rendering different checklist.
Did you take that into account?

capJavert replied:
Yeah, steps creation is completely optional and can be what ever, this is just an example of one implementation.

capJavert replied:
I'll refactor to show concrete example for the spec.

capJavert replied:
@rebelchris see `SquadChecklistCard`

rebelchris commented on packages/shared/src/styles/base.css:
```diff
@@ -52,6 +52,7 @@ body {
   --theme-background-bun: theme('colors.bun.80');
   --theme-background-onion: theme('colors.onion.80');
   --theme-background-pepper: theme('colors.pepper.20');
+  --theme-background-cabbage-opacity-24: theme('colors.cabbage.50')24;
```
Are you sure we need to introduce a new color, would think it should already exist.

rebelchris reviewed (changes requested):
A generic note, the colors you use are not correct and technically the eslint should pick up on this. Might have to re-evaluate that.

Also noticed you choose per "state+role" checklist, in the last decision we choose a generic checklist, did you consider that case?

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,91 @@
+
+  return (
+    <div className={className}>
+      <Card className="p-0 rounded-14 border-cabbage-40 hover:!border-cabbage-40 w-[340px]">
```
This line still have some wrong colors.

rebelchris replied:
Also make sure you covered the mobile view I think that might be missing from this PR now.

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,96 @@
+  const isOpen = !isCompleted && checked;
+
+  return (
+    <div className={className.root}>
```
Not urgent, but think normally we would call this container

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,96 @@
+        }}
+      >
+        <div className="flex flex-1 gap-1 items-center flex-start">
+          <QuaternaryButton
```
Why do we need the button inside a button?
Can't the outside button carry all the styling directly?

capJavert replied:
I added this just so styles match what is in sketch (since it is that button there) outside one is a actually clickable handler. Let me know what would be better way to do it.

capJavert replied:
I did have styling issues with text wrapping and positioning because `children` of `QuaternaryButton` are wrapped in `label` which also does not have styling exposed. 

I removed the `QuaternaryButton` in fcc3ad5635c3ed36edc3c1d63fca3dc36bb524fa and just used plain styles, it looks the same.

rebelchris commented on packages/shared/src/components/checklist/InstallExtensionChecklistStep.tsx:
```diff
@@ -0,0 +1,25 @@
+    <ChecklistStep {...props}>
+      <Button className="btn-primary">
+        <FlexCentered className="gap-2">
+          <BrowsersIcon className="text-theme-label-primary !w-12" /> Add to
```
This would actually be better as a plain SVG then, it seems weird to have a icon that we by default have to modify.

rebelchris commented on packages/shared/src/components/checklist/NotificationChecklistStep.tsx:
```diff
@@ -0,0 +1,19 @@
+import BellIcon from '../icons/Bell';
+
+const NotificationChecklistStep = (props: ChecklistStepProps): ReactElement => {
+  // TODO WT-1293-checklist-components add business logic
```
Just a question how can we maintain that this happens when we merge this to intermediate?
I want to avoid us forgetting about it.

capJavert replied:
Added tasks already to board.

rebelchris commented on packages/shared/src/components/icons/ChecklistA/index.tsx:
```diff
@@ -0,0 +1,15 @@
+import FilledIcon from './filled.svg';
+import OutlinedIcon from './outlined.svg';
+
+const ChecklistAIcon = ({ size, ...rest }: IconProps): ReactElement => (
```
I'm still a bit confused what this means `ChecklistA` can't we give it a more explicit name?
Pretty sure we are not using `Checklist` anyway so maybe just use that?

capJavert replied:
This is how it was in sketch, thought it would be best to just name them the same so down the line we know which icons we added and which ones we are missing.

rebelchris replied:
Oh I see, would you mind checking with Design if they can rename it? 😅
If not it's fine, but would clean it a bit.

capJavert replied:
I'll just rename it, don't want to waste time on.

rebelchris commented on packages/shared/src/graphql/actions.ts:
```diff
@@ -2,6 +2,13 @@ import request, { gql } from 'graphql-request';
 import { graphqlUrl } from '../lib/config';
 
 export enum ActionType {
+  CreateSquad = 'createSquad',
```
Just make sure these match the backend as I think we use snake_case there.

capJavert replied:
Done

rebelchris commented on packages/shared/src/lib/checklist.ts:
```diff
@@ -0,0 +1,73 @@
+  children?: ReactNode;
+};
+
+export const actionsPerRoleMap: Partial<
```
Not 100% sure this is valid as certain types can show based outside of role (for instance the create my feed.
Would this cover that?

capJavert replied:
But checklist is always firstly shown based on squad role, and then we would just add that action to both roles.

rebelchris commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,91 @@
+  return (
+    <div className={className}>
+      <Card className="p-0 rounded-14 border-cabbage-40 hover:!border-cabbage-40 w-[340px]">
+        <div className="p-4 rounded-t-14 bg-theme-bg-cabbage-opacity-24">
```
Is this not `cabbage-opacity-24`?

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,96 @@
+            icon={
+              <div
+                className={classNames(
+                  active && 'p-1 rounded-full bg-theme-bg-cabbage-opacity-24',
```
Same here is this one not `cabbage-opacity-24`?

capJavert replied:
I do not see this class existing

rebelchris replied:
That the new one you introduced in this PR no?

capJavert replied:
Yeah, it is called `bg-theme-bg-cabbage-opacity-24`, do you want me to rename or?

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,96 @@
+  children,
+}: ChecklistStepProps): ReactElement => {
+  const isCompleted = !!step.action.completedAt;
+  const isOpen = !isCompleted && checked;
```
Does this one mean every non completed step is open?
I think the idea is that only one step at the time will be active.

capJavert replied:
No, just a clunky naming, refactored in 9409c10e30a4dd22875ead5299d74669e3276f03

rebelchris reviewed (changes requested):
Wow really looking great already.
Have some suggestions and questions around certain aspects.

capJavert commented:
@rebelchris I think I fixed everything from the comments and also refactored some stuff for easier reuse. Also added hide/show logic and mobile style (modal). Aside from some design stuff pending this is ready for review again.

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,95 @@
+  const isCompleted = !!step.action.completedAt;
+
+  return (
+    <Container className={className.root}>
```
Ah sorry think you misinterpreted my previous comment, I mean instead of `root` we use `container` normally. Not specifically the div :)

rebelchris commented on packages/shared/src/components/checklist/ChecklistStep.tsx:
```diff
@@ -0,0 +1,95 @@
+
+  return (
+    <Container className={className.root}>
+      <button
```
Way to go on. this! Looks way neater with just the one button.

rebelchris commented on packages/shared/src/components/checklist/SquadChecklistCard.tsx:
```diff
@@ -0,0 +1,59 @@
+    return null;
+  }
+
+  const checklistElement = (
```
would it make any difference defining this after the modal return as it's not used if that's the case?

capJavert replied:
It is used in both cases. I just use Modal vs InteractivePopup as wrappers.

rebelchris commented on packages/shared/src/components/squads/SquadHeaderBar.tsx:
```diff
@@ -47,6 +50,26 @@ export function SquadHeaderBar({
+  const { steps } = useSquadChecklistSteps({ squad });
+
+  const checklistTooltipText = useMemo(() => {
+    const completedStepsCount = steps.filter(
```
I have another suggestion here.
Maybe you can abstract the map inside the hook, so it's outside the memo.
Then keep this logic inside the hook (like `completedSteps`)
raising it because down the line you might want it somewhere else then all logic is handled in the hook.

You could also add `totalSteps` for instance

Let me know what you think.

capJavert replied:
We can move this to `useSquadChecklistSteps` sure 👍

capJavert replied:
Streamlined usage even more.

rebelchris commented on packages/shared/src/components/squads/SquadHeaderMenu.tsx:
```diff
@@ -65,12 +66,10 @@ export default function SquadHeaderMenu({
-            type: LazyModal.SquadTour,
-          }),
-        label: 'Learn how Squads work',
+        Icon: FeedbackIcon,
```
Did you remove the tour in favor, not 100% sure that was intended (but could be wrong)
Please just make double sure they don't just want to add a option.

capJavert replied:
Per design it should be removed but I'll check.

rebelchris commented on packages/shared/src/components/tooltips/ContextMenuItem.tsx:
```diff
@@ -17,14 +18,17 @@ function ContextMenuItem({
     <ConditionalWrapper
       condition={!!href}
       wrapper={(children) => (
         <Link href={href}>
```
Not sure don't you need target here as well for nextjs?

capJavert replied:
That should work on newer versions of Next.js. This one needs `target` on `a` to work. Tested it.

rebelchris commented on packages/shared/src/hooks/useChecklist.ts:
```diff
@@ -0,0 +1,51 @@
+};
+
+const useChecklist = ({ steps }: UseChecklistProps): UseChecklist => {
+  const activeStep = useMemo(
```
Not 100% sure, but maybe it's also something for the hook?
In my opinion all calculation regarding to steps should happen in that isolated scoped, but let me know what you think.  (pros/cons)

capJavert replied:
Actually my main point here is that I wish to separate base checklist logic vs specific steps logic. 

In the future we will have other checklist for new features that we don't even know about and this is where that kind of separation will hopefully pay out. 

For example active step for the checklist will be the same always, it is the next uncompleted step in steps list and it is defined by spec/design. Checklist does not know how many more steps are there or what they do, it just knows it needs to show active/next step to the user. 

This is also why each `ChecklistStep` can accept `component` where you can customize what each step does on per case by case basis.

So to conclude `useChecklist` and `Checklist*` components contains all the generic checklist logic that is tied to checklist UI and then steps creation is something that will be done separately for each new checklist eg. for squads we have `useSquadChecklistSteps` and `SquadChecklistCard` where that specific logic lives.

Hope it makes more sense.

rebelchris replied:
Ah yes got it, so in that case we do need to ensure all calculations happen from the relevant hook, not the UI endpoints correct? (so useSquadChecklist)

capJavert replied:
Yep, I streamlined everything even more in latest commit. Same as composing `Checklist*` components into `SquadChecklistCard` I did the same with `useChecklist` hook inside `useSquadChecklist`.

rebelchris commented on packages/webapp/pages/squads/[handle]/index.tsx:
```diff
@@ -62,6 +62,14 @@ const SquadEmptyScreen = dynamic(
     ),
 );
 
+const SquadChecklistCard = dynamic(async () => {
```
Just for my understanding why did you choose dynamic import when it's always rendered?
Seems to have no specific benefit that way.

rebelchris replied:
Alright cool makes sense though was just wondering the thoughts.
I'm not against it especially if it has benefits 🙏

rebelchris reviewed (changes requested):
Got some minor suggestions, but looking very good so far :)

capJavert commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,75 @@
+import { useChecklist } from '../../hooks/useChecklist';
+import { ButtonSize } from '../buttons/Button';
+import CloseButton from '../CloseButton';
+import { RankConfetti } from '../../svg/RankConfetti';
```
Used existing confetti.

capJavert commented:
@sshanzel I would appreciate if you could also take a look just so we are all aligned.

rebelchris commented on packages/shared/src/components/squads/SquadHeaderBar.tsx:
```diff
@@ -88,14 +95,23 @@ export function SquadHeaderBar({
         />
       )}
-      <SimpleTooltip placement="top" content="Feedback">
+      <SimpleTooltip
```
Totally fine for now, just noting down if this file is getting bloated maybe make a new component for this particular item. (non blocking/non urgent)

rebelchris commented on packages/shared/src/graphql/actions.ts:
```diff
@@ -2,6 +2,14 @@ import request, { gql } from 'graphql-request';
+  MyFeed = 'my_feed',
+  SquadInvite = 'squad_invite',
+  BrowserExtension = 'browser_extension',
   Notification = 'notification',
```
This one is different though, make sure it's aligned with the DR

capJavert replied:
It is already used as this in the PR that @sshanzel merged. @sshanzel maybe you can look if this should be `enable_notification`, I might have missed something.

sshanzel replied:
Right. That is the placeholder value. We should have it converted to `enable_notification`. We haven't released this to prod (app.daily.dev) so it is safe for us to update here.

rebelchris commented on packages/shared/src/hooks/useChecklist.ts:
```diff
@@ -0,0 +1,59 @@
+    () => steps.find((item) => !item.action.completedAt)?.action.type,
+    [steps],
+  );
+  const [openStep, setOpenStep] = useState<string>(activeStep);
+
+  useEffect(() => {
+    setOpenStep(activeStep);
+  }, [activeStep]);
```
Why do you need both?
Is it not enough to simply use `activeStep`?

capJavert replied:
This is to sync up when `activeStep` changes due to next action being done, it will auto focus next action in the checklist. It actually provides good UX.

rebelchris commented on packages/shared/src/hooks/useChecklist.ts:
```diff
@@ -0,0 +1,59 @@
+    });
+  }, []);
+
+  const isDone = useMemo(() => {
+    return steps.every((item) => !!item.action.completedAt);
+  }, [steps]);
+
+  const completedSteps = useMemo(() => {
+    return steps.filter((step) => !!step.action.completedAt);
+  }, [steps]);
```
Do you need useMemo for each individual item if you already memoize the whole return object? (Not 100% sure)
Also tagging @idoshamun from his point of view.

capJavert replied:
We would not need it if we compute everything inside that single `useMemo`, if not `completedSteps` would for example have new reference on each render and then again useMemo at the bottom would recompute on each render making it invalid. The new rule of hooks would actually catch that in this case.

In regards to `useMemo` at the bottom in return, all of our hooks return stable reference so at this point just a consistency.

rebelchris replied:
Not sure they would get new reference if steps doesn't change?
I personally think just running useMemo on the return would be enough. *(this just feels like a overkill at this point)

sshanzel replied:
I think the one creating new references is the `.filter`.

capJavert replied:
To be honest I would prefer removing return object `useMemo` from all of our hooks since we rarely need reference equality for the whole return value and most often use destructuring for specific properties, which is better pattern of course. Then it is also easier to selectively memoize specific values. 

I am open on removing return object `useMemo` for `useChecklist` and also we can add it as a rule of the thumb for all new hooks.

capJavert replied:
`useMemo` does not control re-renders of the component that is using it. It only optimizes potentially heavy computations of the values it accepts. Having or not having `useMemo` component will re-render if:
1. its state changes (eg. useState)
2. any of the parent props that component (and/or hook) accepts changes

With `useMemo` we by default merely save on computations during those re-renders. Take the above `completedSteps` calculation. Without `useMemo` it would run on every render hence increasing render time. With `useMemo` it will run only if `steps` changes and will use memoized value on every other re-render (skipping actual iteration in `filter`).

Where `useMemo` helps potentially is with re-renders of children components. By default React function components will always re-render if parent component re-renders, which as we said above will happen due to points 1. or 2. Even if we memoize our props children components will re-render if parent component re-renders. This is how React works.

Where we can fruit benefits is if we were to use [memo](https://react.dev/reference/react/memo) on our components. This wrapper will skip rendering components that accept primitive values as props if neither of them change. `memo` also accepts `areEqual` function which can be used to handle non primitive values but much better way of handling this is to just always pass primitive values. How can we do that with arrays and objects who obviously are not primitives and they reference change? We `useMemo` to make them primitive. 

To go back to our `completedSteps` example which is `array` and not primitive value. Without `useMemo` it would have a new reference on each render and our `memo` wrapped component would re-render because it does not know it is actually visually same array. React under the hood uses [Object.is](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is) to check that equality. 

With `useMemo` reference will stay the same and `memo` will see it as primitive value. It will only change reference if its dependencies change (in this case `steps`) and only then it will also trigger re-render of our `memo` wrapped children component. This is where we would directly save on re-renders with `useMemo`.

This is why removing `useMemo` from return object would no really do anything unless we were to pass whole return object to some component that uses `memo`. Most of the time we pass specific properties of our return object as props to our children components and this is why I mentioned below statement:

> To be honest I would prefer removing return object useMemo from all of our hooks since we rarely need reference equality for the whole return value and most often use destructuring for specific properties, which is better pattern of course. Then it is also easier to selectively memoize specific values.

In all other cases I like to think about `useMemo` as just a in render optimisation that can potentially (if needed) lead to easier re-render optimisations if we ever need to introduce `memo` on some of our components.

Some additional points:
1. `useMemo` of course helps with `useEffect` which also only knows how to check primitive values, and if `useEffect` calls a state setter then we could say `useMemo` here helped with re-renders eg. if we had `useEffect` that was dependant of our `completedSteps`
2. Should we add `memo` to all of our components, not really, even though it does a lot of work (vs some other frameworks like Svelte and SolidJS) React is plenty fast 99% of the time, only when we see some bottlenecks then we can utilise `memo`
3. We also use some 3party packages and components which do some complex stuff might be wrapped in `memo` by library author so by using `useMemo` we potentially save there
4. There is also a big discussion of `useMemo` overhead vs just relying on V8 runtime optimizations. My rule of the thumb that I sleep better if any kind of iteration or algorithmic function inside components is behind `useMemo`.. but that is a rabbit whole for another day 😆 

I hope I explained it here, would love to talk more about this if needed any time.

capJavert replied:
@idoshamun @rebelchris @sshanzel since we are merging this to epic branch and we are onboard for all the features part of the PR I'll merge this now so we can continue on other squads v5 tasks and I can do another PR to resolve `useMemo` depending on the path we decide to take.

idoshamun replied:
Happy to continue the discussion here even though you merged :)
So one thing that you missed is that whenever you define a new return object as json, even if the underlying props are the same it will be referred as a new object and will trigger re-render for those who use this hook. I encountered few years ago when I saw a suggestion to useMemo when returning a json object. Again my assumption might have changed but this is another use case for useMemo

capJavert replied:
> So one thing that you missed is that whenever you define a new return object as json, even if the underlying props are the same it will be referred as a new object and will trigger re-render for those who use this hook.

Yes, this is true, but it would need to be used in specific ways, let's look at specific examples on when would this matter:
1. The whole hook value needs to be passed as dependency so:
```jsx
const checklist = useChecklist()

return <Child checklist={checklist} />
```

If `Child` is component wrapped in `memo` it will now re-render every time if `useChecklist` return object is not using `useMemo` because of course object is not primitive value. This is where return object `useMemo` matters as you said. ✅ 

That said, we mostly have this kind of usages:
```jsx
const checklist = useChecklist()

return <Child {...checklist} />
```

In above example it does not matter if return object is memoized or not it only matters if specific properties are memoized (and they are in the case of `useChecklist`) eg. `completedSteps`, `isDone`...

The above is of course shortcut for something like this:
```jsx
const { completedSteps, isDone } = useChecklist()

return <Child completedSteps={completedSteps} isDone={isDone}  />
```

The same principal applies here, `useMemo` is not needed for return object if we already memoize `completedSteps` and `isDone` separately, since those are actually passed as props separately.

Of course this all comes with a caveat that we have to use `memo` for `Child`, if it is just a function component it will render no matter no matter the props if its parent component(s) re-render.

Your point also stands for `useEffect` where of course we have this:
```jsx
const checklist = useChecklist()

useEffect(() => {
    // do something
}, [checklist])
```

it matters that return object is memoized here since it is directly used as dependency.

Again, we mostly use stuff like this.

```jsx
const { completedSteps } = useChecklist()

useEffect(() => {
    // do something
}, [completedSteps])
```

This is why I think according of how we use our special hooks we do not need to do this double memoization (for return object and separate properties), but leaving it we are not loosing anything drastic and re-renders wise we are not really saving anything since we do not utilise `React.memo` wrapper.

idoshamun replied:
I'd love to see a minimal project where we can compare both and see the number of re-renders. You might be right and if so we need to have proper guidelines on that

capJavert replied:
Here is a minimal implementation to see how different examples above re-render https://codesandbox.io/s/re-render-tests-with-usememo-and-memo-p7dg9r (see `index.js` for info on how to use different examples)

idoshamun replied:
Thanks for setting it up so quickly. I think we need a bit different example, we need to check what happens when `steps` is updated by the deep value is the same. Whether it will cause re-renders in some cases or not. Anyway, I think it will be better to have a meeting all of us :)
Feel free to schedule.

rebelchris reviewed (approved):
Looks pretty neat!
Like the hooks you created, just have some minor non blocking comments.

Will approve from my side, keep in mind we need 2 approvals for this one as it's quite big.

sshanzel commented on packages/shared/src/styles/base.css:
```diff
@@ -166,6 +167,7 @@ body {
   --theme-background-bun: theme('colors.bun.20');
   --theme-background-onion: theme('colors.onion.20');
   --theme-background-pepper: theme('colors.pepper.20');
+  --theme-background-cabbage-opacity-24: theme('colors.cabbage.50')24;
```
Not sure if I am looking at the right Sketch part relevant to the usage of this.

The spec says 40 and opacity of 18%.

`18%` opacity I think should be `2B`.

Link for reference: https://gist.github.com/lopspower/03fb1cc0ac9f32ef38f4

![Screenshot 2023-05-04 at 8 22 37 PM](https://user-images.githubusercontent.com/13744167/236202898-c40b79f5-d232-4ad7-871b-5de22c1a97dc.png)

capJavert replied:
Checked with Tomer, it should be `24%` to match rest of the elements.

sshanzel commented on packages/shared/src/hooks/useActions.ts:
```diff
@@ -21,7 +21,7 @@ export const useActions = (): UseActions => {
     generateQueryKey(RequestKey.Actions, user),
     getUserActions,
-    { enabled: !!user },
+    { enabled: !!user?.id },
```
This change shouldn't be needed as it is coming from the context, and if it is anonymous user, we return null for user.

capJavert replied:
I just thought no point in doing a separate API request for anonymous since it will just return nothing. This way it should be only one once we know user is logged in and actually capable of completing actions.

sshanzel replied:
I mean on an anonymous user this won't make a request still because the `user` object is null if it is anonymous.

capJavert replied:
In some cases I was getting `{ name: 'anonymous' }` or something like that, tried it now and I can not get it anymore. You might know?

sshanzel replied:
Oh, interesting. I'll be sure to watch out if I encounter any. Just FYI, if ever you need the anonymous data (like tracking id) there should be an `anonymous` property from `AuthContext` 🚀

sshanzel commented on packages/shared/src/components/tooltips/ContextMenuItem.tsx:
```diff
@@ -1,14 +1,15 @@
   href?: string;
   label: string;
   Icon: React.ComponentType<IconProps>;
+  target?: HTMLAttributeAnchorTarget;
```
As the props grow, I think it would be best to have the whole anchor attributes in a single property as we discussed in the past, then just pass along whatever is needed.

capJavert replied:
Added in 030a5a0ccb4cc68bd5fbdb828c440306e8fa5884

sshanzel commented on packages/shared/src/components/checklist/ChecklistCard.tsx:
```diff
@@ -0,0 +1,75 @@
+  return (
+    <div className={className}>
+      <Card className="p-0 bg-transparent rounded-14 !border-theme-color-cabbage hover:!border-theme-color-cabbage w-[340px]">
+        <div className="overflow-hidden relative p-4 bg-gradient-to-t from-cabbage-90 to-cabbage-50 rounded-t-12">
```
Not sure we should be using `from-cabbage-90 to-cabbage-50`. We need to use the one defined in our config, we might just have to find the correct one. I understand it is a pain for now, but hopefully we can create a documentation mapping this.

rebelchris replied:
I think for this it's fine it's a exclusion that's the same render for both light/dark modes 😅
It's a bit tricky but we used it before like this. (Correct me if i'm wrong though)

sshanzel replied:
Ah. On the banner. We might probably need to come up with some sort of system for this in the future. Just to be consistent and inline. In that case, kindly disregards this one 🚢

capJavert replied:
I see I posted comment late without seeing your comments. 

Ok, going with this, tnx for checking 🍻

rebelchris replied:
Agree with Lee though, @capJavert can you add a backlog ticket just so we have it :)

capJavert replied:
For sure!

capJavert replied:
https://dailydotdev.atlassian.net/browse/WT-1316

sshanzel reviewed (changes requested):
Wow! Massive work here 💪 

Really amazing with all the thoughts put into this, just some minor comments.

sshanzel commented on packages/shared/src/graphql/actions.ts:
```diff
@@ -2,7 +2,15 @@ import request, { gql } from 'graphql-request';
+  MyFeed = 'my_feed',
+  SquadInvite = 'squad_invite',
+  BrowserExtension = 'browser_extension',
+  EnableNotification = 'enable_notification',
```
Thanks for fixing this!

sshanzel commented on packages/shared/src/styles/base.css:
```diff
@@ -52,6 +52,7 @@ body {
   --theme-background-bun: theme('colors.bun.80');
   --theme-background-onion: theme('colors.onion.80');
   --theme-background-pepper: theme('colors.pepper.20');
+  --theme-background-cabbage-opacity-24: theme('colors.cabbage.40')24;
```
The hex value for 24% should be 3D, I think.

Reference: https://gist.github.com/lopspower/03fb1cc0ac9f32ef38f4

```suggestion
  --theme-background-cabbage-opacity-24: theme('colors.cabbage.40')3D;
```

capJavert replied:
Did not realize we are using hex values. Updated 👌

sshanzel replied:
Me too, it was until @rebelchris gave me that link months ago 😂

sshanzel commented on packages/shared/src/styles/base.css:
```diff
@@ -166,6 +167,7 @@ body {
   --theme-background-bun: theme('colors.bun.20');
   --theme-background-onion: theme('colors.onion.20');
   --theme-background-pepper: theme('colors.pepper.20');
+  --theme-background-cabbage-opacity-24: theme('colors.cabbage.40')24;
```
Just tagging for visibility.
```suggestion
  --theme-background-cabbage-opacity-24: theme('colors.cabbage.40')3D;
```

sshanzel reviewed (approved):
Got nothing else to say but great work! 🚢 🚢 🚢
